src/vm/prestub.cpp

```diff
@@ -83,58 +82,113 @@ PCODE MethodDesc::DoBackpatch(MethodTable * pMT, MethodTable *pDispatchingMT, BO
     {
         STANDARD_VM_CHECK;
         PRECONDITION(!ContainsGenericVariables());
-        PRECONDITION(HasStableEntryPoint());
         PRECONDITION(pMT == GetMethodTable());
     }
     CONTRACTL_END;
-    PCODE pTarget = GetStableEntryPoint();
 
-    if (!HasTemporaryEntryPoint())
-        return pTarget;
+    bool isTieredVtableMethod = IsTieredVtableMethod();
+    LoaderAllocator *mdLoaderAllocator = isTieredVtableMethod ? GetLoaderAllocator() : nullptr;
+
+    // Only take the lock if it's a tiered virtual method, for recording slots and synchronizing with backpatching slots
+    MethodDescBackpatchInfoTracker::ConditionalLockHolder lockHolder(isTieredVtableMethod);
+
+    // For tiered vtable methods, get the method entry point inside the lock above to synchronize with backpatching in
+    // MethodDesc::BackpatchEntryPointSlots()
+    PCODE pTarget = GetMethodEntryPoint();
+
+    PCODE pExpected;
+    if (isTieredVtableMethod)
+    {
+        _ASSERTE(pTarget == GetEntryPointToBackpatch_Locked());
+
+        pExpected = GetTemporaryEntryPoint();
+        if (pExpected == pTarget)
+            return pTarget;
 
-    PCODE pExpected = GetTemporaryEntryPoint();
+        // True interface methods are never backpatched and do not qualify for tiered vtable slot backpatching
+        _ASSERTE(!(pMT->IsInterface() && !IsStatic()));
```

Comment:
I'm not sure why it would have been there before, but when I tried to enable it I saw an issue related to default interface methods and method impls. It looked like backpatching a method impl slot on an interface type can be invalid. IIRC it was the [diamondshape test](https://github.com/dotnet/coreclr/tree/master/tests/src/Loader/classloader/DefaultInterfaceMethods/diamondshape) that failed due to calling a wrong method as a result of incorrect backpatching but I don't remember the details now.